// scrypt-ts/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.TestWallet = exports.Provider = exports.replayToLatest = exports.bsv = exports.FunctionCall = exports.buildOpreturnScript = exports.buildPublicKeyHashScript = exports.toHex = exports.SmartContractLib = exports.SmartContract = exports.method = exports.prop = exports.P2PK = exports.P2PKH = void 0;
const decorators_1 = require("./smart-contract/decorators");
Object.defineProperty(exports, "prop", { enumerable: true, get: function () { return decorators_1.prop; } });
Object.defineProperty(exports, "method", { enumerable: true, get: function () { return decorators_1.method; } });
const contract_1 = require("./smart-contract/contract");
Object.defineProperty(exports, "SmartContract", { enumerable: true, get: function () { return contract_1.SmartContract; } });
const library_1 = require("./smart-contract/library");
Object.defineProperty(exports, "SmartContractLib", { enumerable: true, get: function () { return library_1.SmartContractLib; } });
var p2pkh_1 = require("./smart-contract/builtins/p2pkh");
Object.defineProperty(exports, "P2PKH", { enumerable: true, get: function () { return p2pkh_1.P2PKH; } });
var p2pk_1 = require("./smart-contract/builtins/p2pk");
Object.defineProperty(exports, "P2PK", { enumerable: true, get: function () { return p2pk_1.P2PK; } });
__exportStar(require("./smart-contract/builtins/types"), exports);
__exportStar(require("./smart-contract/builtins/functions"), exports);
__exportStar(require("./bsv/utils"), exports);
__exportStar(require("./smart-contract/utils"), exports);
var scryptlib_1 = require("scryptlib");
Object.defineProperty(exports, "toHex", { enumerable: true, get: function () { return scryptlib_1.toHex; } });
Object.defineProperty(exports, "buildPublicKeyHashScript", { enumerable: true, get: function () { return scryptlib_1.buildPublicKeyHashScript; } });
Object.defineProperty(exports, "buildOpreturnScript", { enumerable: true, get: function () { return scryptlib_1.buildOpreturnScript; } });
Object.defineProperty(exports, "FunctionCall", { enumerable: true, get: function () { return scryptlib_1.FunctionCall; } });
var index_1 = require("./smart-contract/bsv/index");
Object.defineProperty(exports, "bsv", { enumerable: true, get: function () { return index_1.bsv; } });
__exportStar(require("./bsv/types"), exports);
__exportStar(require("./smart-contract/types/index"), exports);
__exportStar(require("./smart-contract/utils/index"), exports);
var replay_1 = require("./smart-contract/replay");
Object.defineProperty(exports, "replayToLatest", { enumerable: true, get: function () { return replay_1.replayToLatest; } });
/* export Providers */
__exportStar(require("./providers"), exports);
var abstract_provider_1 = require("./bsv/abstract-provider");
Object.defineProperty(exports, "Provider", { enumerable: true, get: function () { return abstract_provider_1.Provider; } });
/* export Wallets */
__exportStar(require("./bsv/abstract-signer"), exports);
var test_wallet_1 = require("./bsv/wallets/test-wallet");
Object.defineProperty(exports, "TestWallet", { enumerable: true, get: function () { return test_wallet_1.TestWallet; } });
__exportStar(require("./bsv/signers"), exports);
__exportStar(require("./client"), exports);
